engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/LazyAtomInstanceNode.java

```diff
@@ -0,0 +1,163 @@
+package org.enso.interpreter.node.expression.builtin.meta;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.ArityException;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.nodes.RootNode;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.argument.ArgumentDefinition;
+import org.enso.interpreter.runtime.callable.argument.CallArgumentInfo;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.callable.function.FunctionSchema;
+import org.enso.interpreter.runtime.data.Array;
+import org.enso.interpreter.runtime.data.Vector;
+import org.enso.interpreter.runtime.error.PanicException;
+
+@BuiltinMethod(
+    type = "Meta",
+    name = "lazy_atom",
+    description = "Creates a new atom with given constructor and fields.",
+    autoRegister = false)
+public abstract class LazyAtomInstanceNode extends Node {
+
+  static LazyAtomInstanceNode build() {
+    return LazyAtomInstanceNodeGen.create();
+  }
+
+  abstract Vector execute(Object factory);
+
+  @Specialization(guards = "iop.isExecutable(factory)")
+  Vector doExecute(
+    Object factory,
+    @CachedLibrary(limit="3") InteropLibrary iop,
+    @Cached SwapAtomFieldNode swapNode
```

Comment:
This caching structure is strange, with a global (until splitting, if it kicks in) cache that becomes megamorphic (with a truffle boundary!) on first miss. Meanwhile, we could specialize on `factory` being a function and specialize on its call target – this way we get a neat cache in _all_ cases we currently have in mind (where the lambda is just a trivial `Atom arg1 _ arg3`)

Reply:
I don't think I get the problem. The classical usage is [shown here](https://github.com/enso-org/enso/pull/4023/files#diff-e409cbed47b435634ceafa6dab6bba509fc11d47bbf5ca72c1accfb42e9684ecR217) - e.g.:
```
Meta.atom_with_hole (e -> Cons v e)
```
after parsing there will be one `AtomWithAHoleNode` with one `SwapAtomFieldNode`. The `SwapAtomFieldNode` will cache the constant `FunctionSchema` and will speculate on the `lastIndex` - almost all the time `lastIndex` is going to be constant as well.

> all cases we currently have in mind (where the lambda is just a trivial Atom arg1 _ arg3)

Right now I believe the current system works for this case well. I'll dig into IGV now.

Reply:
It is still slower - reported as [pivotal](https://www.pivotaltracker.com/story/show/184171846), but I'd like to integrate anyway - the functional improvements in `List.enso` worth it, I believe.